### PR TITLE
subscriber api endpoint

### PR DESCRIPTION
**Pull request by remstone7**
It appears we don't get customers who subscriber as a guest as the hooks don't fire.  this is to expose 2 api endpoints to be able to query this data.  

1. for historical sync by id's (batching)
2. for periodic date pull (since param)

Endpoints:

```
@param mixed $start
@param mixed $until
@param mixed $store_id
```
/V1/klaviyo/reclaim/customers/subscriptions/periodic
Ex
```
rest/V1/klaviyo/reclaim/customers/subscriptions/periodic?start=2019-04-25%20:00:00&until=2019-04-25%2018:00:00:00
```

```
@param mixed $start_id	   
@param mixed $end_id	     
@param mixed $store_id
```

/V1/klaviyo/reclaim/customers/subscriptions/historical

```
http://magento2.local:8888/rest/V1/klaviyo/reclaim/customers/subscriptions/historical?start_id=1&end_id=10
```

### CODE REVIEW

File: Model/Reclaim.php [hunk 1]
```diff
@@ -127,6 +136,84 @@ public function productinspector($start_id, $end_id){
...
+        if (!$start_id || !$end_id ){ 
+            throw new NotFoundException(__('Please provide start_id and end_id'));
+        }
+
+        $storeIdFilter = $this->_storeFilter($storeId);
+
+        $subscriberCollection =$this->subscriberCollection->create()
+            ->addFieldToFilter('subscriber_id', ['gteq' => (int)$start_id])
+            ->addFieldToFilter('subscriber_id', ['lteq' => (int)$end_id])
+            ->addFieldToFilter('store_id', [$storeIdFilter => $storeId]);
```
Comment (klaviyojad) [inline Model/Reclaim.php]: do we still wanna do that if $storeId is null?

File: Model/Reclaim.php [hunk 2]
```diff
@@ -127,6 +136,84 @@ public function productinspector($start_id, $end_id){
         return $response;
 
     }
+
+    public function historicalcustomersubscription($start_id, $end_id, $storeId=null)
+    {
+
+        if (!$start_id || !$end_id ){
```
Comment (klaviyojad) [inline Model/Reclaim.php]: can we also check that $start_id <$end_id?

File: Model/Reclaim.php [hunk 3]
```diff
@@ -127,6 +136,84 @@ public function productinspector($start_id, $end_id){
...
+        }
+
+        $storeIdFilter = $this->_storeFilter($storeId);
+
+        $subscriberCollection =$this->subscriberCollection->create()
+            ->addFieldToFilter('subscriber_id', ['gteq' => (int)$start_id])
+            ->addFieldToFilter('subscriber_id', ['lteq' => (int)$end_id])
+            ->addFieldToFilter('store_id', [$storeIdFilter => $storeId]);
+
+        $subscriberCollection =$this->subscriberCollection->create()
```
Comment (klaviyojad) [inline Model/Reclaim.php]: why do we do that after getting the start_id to end_id range?

File: Model/Reclaim.php [hunk 4]
```diff
@@ -127,6 +136,84 @@ public function productinspector($start_id, $end_id){
...
+        $response = $this->_packageSubscribers($subscriberCollection);
+
+        return $response;
+    }
+
+    public function customersubscription($start, $until, $storeId=null)
+    {
+        
+        if (!$start || !$until ){ 
+            throw new NotFoundException(__('Please provide start_id and end_id'));
```
Comment (klaviyojad) [inline Model/Reclaim.php]: its different here

**Comment by klaviyojad** [comment]
I dont get the naming convention you use, periodic vs historical?  Why not just call it getSubscribedsById and getSubscribersByDateRange?

File: Model/Reclaim.php (hunk 1 above)
Comment (remstone7) [reply Model/Reclaim.php]: it queries for nlike storeId = null.. was basically to save some additional logic of dropping the 3rd query

File: Model/Reclaim.php (hunk 3 above)
Comment (remstone7) [reply Model/Reclaim.php]: oops, removing from testing

File: Model/Reclaim.php [hunk 5]
```diff
@@ -12,20 +12,29 @@ class Reclaim implements ReclaimInterface
...
         \Magento\Quote\Model\QuoteFactory $quoteFactory,
+        \Magento\Newsletter\Model\Subscriber $subscriber,
+        \Magento\Newsletter\Model\ResourceModel\Subscriber\CollectionFactory $subscriberCollection,
         \Klaviyo\Reclaim\Helper\Data $klaviyoHelper
         )
     {
         $this->quoteFactory = $quoteFactory;
         $this->_objectManager = $objectManager;
+        $this->_subscriber= $subscriber;
+        $this->subscriberCollection = $subscriberCollection;
```
Comment (remstone7) [inline Model/Reclaim.php]: should `_` this

File: Model/Reclaim.php [hunk 6]
```diff
@@ -127,6 +136,81 @@ public function productinspector($start_id, $end_id){
...
+        // $start = '2019-04-25 00:00:00';
+
+        // don't want any big queries, we limit to 10 days
+        $until_date = strtotime($until);
+        $start_date = strtotime($start);
+        $datediff = $until_date - $start_date;
+
+        if (abs(round($datediff / (60 * 60 * 24))) > self::MAX_QUERY_DAYS){
+            throw new NotFoundException(__('Please don\'t query for more than 10 days'));
+        }
```
Comment (invertedfjord) [inline Model/Reclaim.php]: maybe change the wording on this to something like "Cannot query more than 10 days".

File: Model/Reclaim.php [hunk 7]
```diff
@@ -127,6 +136,81 @@ public function productinspector($start_id, $end_id){
...
+        return $response;
+    }
+
+    public function getSubscribersByDateRange($start, $until, $storeId=null)
+    {
+        
+        if (!$start || !$until ){ 
+            throw new NotFoundException(__('Please provide start and until param'));
+        }
+        // start and until date formats
```
Comment (invertedfjord) [inline Model/Reclaim.php]: should we validate date formats as well?

File: etc/webapi.xml [hunk 8]
```diff
@@ -18,6 +18,18 @@
...
+            <resource ref="anonymous"/>
+        </resources>
+    </route>
+    <route url="/V1/klaviyo/reclaim/customers/subscriptions/historical" method="GET">
+        <service class="Klaviyo\Reclaim\Api\ReclaimInterface" method="getSubscribersById"/>
+        <resources>
+            <resource ref="anonymous"/>
+        </resources>
+    </route>
     <route url="/V1/klaviyo/reclaim/" method="GET">
```
Comment (invertedfjord) [inline etc/webapi.xml]: Do we want to add authentication. Doesn't seem too likely, but theoretically we could grab customers subscribers.

File: Model/Reclaim.php [hunk 9]
```diff
@@ -12,20 +12,29 @@ class Reclaim implements ReclaimInterface
...
         $this->quoteFactory = $quoteFactory;
         $this->_objectManager = $objectManager;
+        $this->_subscriber= $subscriber;
+        $this->subscriberCollection = $subscriberCollection;
         $this->_klaviyoHelper = $klaviyoHelper;
 
     }
 
+
     /**
```
Comment (invertedfjord) [inline Model/Reclaim.php]: Nit. remove extra line.

File: etc/webapi.xml (hunk 8 above)
Comment (remstone7) [reply etc/webapi.xml]: yea adding for these

**Pull request merged**